libs/wingc/src/fold.rs

```diff
@@ -0,0 +1,438 @@
+		},
+		signature: f.fold_function_signature(node.signature),
+		is_static: node.is_static,
+		span: node.span,
+		captures: node.captures,
+	}
+}
+
+pub fn fold_function_signature<F>(f: &mut F, node: FunctionSignature) -> FunctionSignature
+where
+	F: Fold + ?Sized,
+{
+	FunctionSignature {
+		parameters: node
+			.parameters
+			.into_iter()
+			.map(|param| f.fold_function_parameter(param))
+			.collect(),
+		return_type: node.return_type.map(|type_| Box::new(f.fold_type_annotation(*type_))),
+		phase: node.phase,
+	}
+}
+
+pub fn fold_function_parameter<F>(f: &mut F, node: FunctionParameter) -> FunctionParameter
+where
+	F: Fold + ?Sized,
+{
+	FunctionParameter {
+		name: f.fold_symbol(node.name),
+		type_annotation: f.fold_type_annotation(node.type_annotation),
+		reassignable: node.reassignable,
+	}
+}
+
+pub fn fold_args<F>(f: &mut F, node: ArgList) -> ArgList
+where
+	F: Fold + ?Sized,
+{
+	ArgList {
+		pos_args: node.pos_args.into_iter().map(|arg| f.fold_expr(arg)).collect(),
+		named_args: node
+			.named_args
+			.into_iter()
+			.map(|(name, arg)| (f.fold_symbol(name), f.fold_expr(arg)))
+			.collect(),
+	}
+}
+
+pub fn fold_type_annotation<F>(f: &mut F, node: TypeAnnotation) -> TypeAnnotation
+where
+	F: Fold + ?Sized,
+{
+	match node {
+		TypeAnnotation::Number => TypeAnnotation::Number,
+		TypeAnnotation::String => TypeAnnotation::String,
+		TypeAnnotation::Bool => TypeAnnotation::Bool,
+		TypeAnnotation::Duration => TypeAnnotation::Duration,
+		TypeAnnotation::Json => TypeAnnotation::Json,
+		TypeAnnotation::MutJson => TypeAnnotation::MutJson,
+		TypeAnnotation::Optional(t) => TypeAnnotation::Optional(Box::new(f.fold_type_annotation(*t))),
+		TypeAnnotation::Array(t) => TypeAnnotation::Array(Box::new(f.fold_type_annotation(*t))),
+		TypeAnnotation::MutArray(t) => TypeAnnotation::MutArray(Box::new(f.fold_type_annotation(*t))),
+		TypeAnnotation::Map(t) => TypeAnnotation::Map(Box::new(f.fold_type_annotation(*t))),
+		TypeAnnotation::MutMap(t) => TypeAnnotation::MutMap(Box::new(f.fold_type_annotation(*t))),
+		TypeAnnotation::Set(t) => TypeAnnotation::Set(Box::new(f.fold_type_annotation(*t))),
+		TypeAnnotation::MutSet(t) => TypeAnnotation::MutSet(Box::new(f.fold_type_annotation(*t))),
+		TypeAnnotation::Function(t) => TypeAnnotation::Function(FunctionTypeAnnotation {
+			param_types: t.param_types.into_iter().map(|t| f.fold_type_annotation(t)).collect(),
+			return_type: t.return_type.map(|t| Box::new(f.fold_type_annotation(*t))),
+			phase: t.phase,
+		}),
+		TypeAnnotation::UserDefined(t) => TypeAnnotation::UserDefined(f.fold_user_defined_type(t)),
+	}
+}
+
+pub fn fold_user_defined_type<F>(f: &mut F, node: UserDefinedType) -> UserDefinedType
+where
+	F: Fold + ?Sized,
+{
+	UserDefinedType {
+		root: f.fold_symbol(node.root),
+		fields: node.fields.into_iter().map(|s| f.fold_symbol(s)).collect(),
+	}
+}
+
+pub fn fold_symbol<F>(_f: &mut F, node: Symbol) -> Symbol
+where
+	F: Fold + ?Sized,
+{
+	node
+}
```

Comment:
This is cool.

My understanding is that this is our mechanism to support syntactic sugar. Correct?

Practically, we will mostly use this to fold expressions to other expressions. Correct?

Reply:
Yep. Statements can probably be folded too if there's a use case for it